<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2022-02-05">

<title>Covid Sequence Clustering â€“ Josh Meehl</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Josh Meehl</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../resume.html"> 
<span class="menu-text">Resume</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-projects" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Projects</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-projects">    
        <li>
    <a class="dropdown-item" href="../projects/protein-engineering.html">
 <span class="dropdown-text">Protein Engineering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/simulated-annealing.html">
 <span class="dropdown-text">Simulated Annealing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/covid-seq-clustering.html">
 <span class="dropdown-text">COVID Strain Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://mage.meehl.org">
 <span class="dropdown-text">The Data Mage</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../contact.html"> 
<span class="menu-text">Contact</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://mage.meehl.org"> 
<span class="menu-text">The Data Mage</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/meehljd/mtg-modeling"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/meehl/"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul class="collapse">
  <li><a href="#the-challenge" id="toc-the-challenge" class="nav-link active" data-scroll-target="#the-challenge">The Challenge</a></li>
  <li><a href="#the-data" id="toc-the-data" class="nav-link" data-scroll-target="#the-data">The Data</a></li>
  <li><a href="#the-method" id="toc-the-method" class="nav-link" data-scroll-target="#the-method">The Method</a></li>
  <li><a href="#the-results" id="toc-the-results" class="nav-link" data-scroll-target="#the-results">The Results</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Covid Sequence Clustering</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 5, 2022</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>In this project, I compare machine learning techniques for clustering of SARS-CoV-2 cDNA sequences by variants.</p>
<p>At this point the whole world has had a crash course in virology, but to recap, the SARS-CoV-2 virus is an RNA virus. These types of viruses often mutate, due to the relative instability of RNA sequences. This results in many strains of the virus, some of which are more infectious than others.</p>
<p>For those who want to dive deeper into the nitty-gritty details, you can find the notebooks in the <a href="https://github.com/meehljd/covid-seqs-clustering">GitHub repository</a>.</p>
<section id="the-challenge" class="level2">
<h2 class="anchored" data-anchor-id="the-challenge">The Challenge</h2>
<p>There are two items we need to address with clustering of cDNA sequences. First, we need to find a way to compare the sequences. Second, we need to find a way to group the sequences. This is a classic unsupervised learning problem.</p>
<p>To challenge myself, I will implement clustering solutions I currently use. Typically I would first do a literature search to find state-of-the-art approaches, but want to solve this problem with my own approaches.</p>
<p>Also, I will avoid libraries like <code>Biopython</code> that are specifically designed for bioinformatics. Instead I will use raw python and <code>numpy</code> to manipulate sequences. For the linear algebra of the clustering techniques, I will use solely <code>numpy</code> and only use <code>sklearn</code> for some convenience functions. Again, this is to challenge myself to implement algorithms from scratch where I would typically use 3rd party libraries. This is reinventing the wheel, but sometimes itâ€™s worth it to get that sense of accomplishment.</p>
</section>
<section id="the-data" class="level2">
<h2 class="anchored" data-anchor-id="the-data">The Data</h2>
<p>The data I have is multi-sequence alignment (MSA) of 400 SARS-CoV-2 sequences in a fasta file. The sequences are of complementary DNA, sequenced from viral RNA that has undergone reverse transcriptase. The sequences are from sample that originate from an unknown number of strains.</p>
<p>Each location of the sequences can have the four nucleotides <code>A</code>, <code>C</code>, <code>G</code>, or <code>T</code>. A position could also have a <code>N</code> if the nucleotide was indeterminate or a <code>-</code> if there was no sequence data at that position. The sequences have been aligned and fully conserved regions replaced with the symbol <code>.</code>.</p>
</section>
<section id="the-method" class="level2">
<h2 class="anchored" data-anchor-id="the-method">The Method</h2>
<p>To cluster the sequences, I first needed to clean the data. Next I needed to find a similarity measure to compare the sequences. Then I used a variety of clustering algorithms to group the sequences.</p>
<section id="data-cleaning" class="level3">
<h3 class="anchored" data-anchor-id="data-cleaning">Data Cleaning</h3>
<p>I used python to parse the fasta file. I then converted the 400 sequences into a <code>numpy</code> array with the characters converted to integers. This gave me a 400 x 34,044 array.</p>
<p>This left me with sequences that contained missing (<code>-</code>) or indeterminate (<code>N</code>) data. If a given MSA position had less that 5% missing data, I imputed the missing data with the most common nucleotide at that position. If a position had more than 5% missing data, I removed the sequence from the dataset. This resulted with the typical sequence having 1-2% of the positions imputed.</p>
<p>At this point I also dropped all fully conserved positions (<code>.</code>), as they would not contribute to the clustering. This left me with a 400 x 17,463 array.</p>
<div id="fig-elephants" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-elephants-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-elephants" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-1" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-width="params.img_max_width">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_head.png" class="lightbox" data-gallery="fig-elephants" title="Figure&nbsp;1&nbsp;(a): Raw MSA"><img src="../images/projects/csc/sequences_head.png" class="img-fluid figure-img" data-max-width="params.img_max_width" data-ref-parent="fig-elephants"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Raw MSA
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-elephants" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-2" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-width="params.img_max_width">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_clean_head.png" class="lightbox" data-gallery="fig-elephants" title="Figure&nbsp;1&nbsp;(b): Cleaned MSA"><img src="../images/projects/csc/sequences_clean_head.png" class="img-fluid figure-img" data-max-width="params.img_max_width" data-ref-parent="fig-elephants"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Cleaned MSA
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-elephants-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Visualization of MSA sequences before and after cleaning, showing first 500 positions. (a) is color scaled to have missing or fully conserved values in black and other values in lighter gray. (b) is rescaled to have A in white and T in black.
</figcaption>
</figure>
</div>
<p>Next I calculated the entropy of the MSA for each position. This gives us an idea of how variable the sequence positions are. We get a sense of baseline noise of random mutations, as well as positions that have low conservation across major strains.</p>
<p>Note for the entropy calculation, I used <span class="math inline">\(log_4\)</span> to account for the number of nucleotides, although hte convention is <span class="math inline">\(log_2\)</span> for bits. To me this makes more sense, since DNA is quaternary, not binary.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="../images/projects/csc/entropy.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Entropy per MSA position. Zoomed to show detail."><img src="../images/projects/csc/entropy.png" class="img-fluid figure-img" data-max-width="params.img_max_width" alt="Entropy per MSA position. Zoomed to show detail."></a></p>
<figcaption>Entropy per MSA position. Zoomed to show detail.</figcaption>
</figure>
</div>
<p>Most positions are fully conserved with an entropy of 0. Some positions have small entropy values, representing positions that are conserved with occasional random mutations. Some positions have high entropy values, representing positions that are highly variable and may indicate mutations across major strains.</p>
<p>Next I plotted a histogram of the entropy values to visualize the random mutations versus the major strain mutations. We can see below the distribution is multimodal. I put a cutoff threshold at 0.145, which is at the right tail of the random mutation distribution. This cutoff will be used to filter out positions that are not conserved across major strains. Future work could look at fitting a mixture model to the entropy distribution to find the optimal cutoff.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="../images/projects/csc/entropy_hist.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Histogram of the entropy per MSA position. The left mode is random mutations and the other modes are mutations of major strains."><img src="../images/projects/csc/entropy_hist.png" class="img-fluid figure-img" data-max-width="params.img_max_width" alt="Histogram of the entropy per MSA position. The left mode is random mutations and the other modes are mutations of major strains."></a></p>
<figcaption>Histogram of the entropy per MSA position. The left mode is random mutations and the other modes are mutations of major strains.</figcaption>
</figure>
</div>
<p>After trimming the MSA to remove low entropy positions, I was left with a 400 x 110 array. This is a good size for clustering, as it is a good balance between having enough data to find clusters and not having too much data to be computationally expensive.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="../images/projects/csc/sequences_trim.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Visualization of MSA sequences after trimming. Color scale based on A (lightest), C, G, T (darkest)."><img src="../images/projects/csc/sequences_trim.png" height="400" alt="Visualization of MSA sequences after trimming. Color scale based on A (lightest), C, G, T (darkest)." class="figure-img"></a></p>
<figcaption>Visualization of MSA sequences after trimming. Color scale based on A (lightest), C, G, T (darkest).</figcaption>
</figure>
</div>
</section>
<section id="similarity-measure" class="level3">
<h3 class="anchored" data-anchor-id="similarity-measure">Similarity Measure</h3>
<p>Now onto the exciting world of similarity measures! The next step in clustering will be to generate a similarity data structure from the sequence data. I will use the Hamming distance as the distance metric. This is a good distance metric for sequences since it works with discrete categorical data. For a pair of sequences, the Hamming distance will be the number of positions where the sequences have different nucleotides.</p>
<p>The trick with Hamming distance is that it gives us a relative similarity between sequence pairs, but not on an absolute scale. We can address this issue by building a graph of the sequences, using the similarity as edge weights. This gives us a good sense of the local structure of the data. We will represent the graph as an adjacency matrix.</p>
<p>To provide cleaner results, I made the dense adjacency matrix sparse. Only the nearest 75 neighbors to a sequence node are retained. The rest are dropped.</p>
</section>
<section id="clustering-algorithms" class="level3">
<h3 class="anchored" data-anchor-id="clustering-algorithms">Clustering Algorithms</h3>
<p>The next step is to cluster the graph. I have two ideas for this, using different perspectives.</p>
<p>The first perspective is to use linear algebra to find a low dimensional representation. We can cluster on the eigenvectors with the lowest eigenvalues. This approach is called <em><a href="https://en.wikipedia.org/wiki/Spectral_clustering">spectral clustering</a></em>.</p>
<p>The second perspective is to treat the neighborhood graph as a low dimensional manifold in high-dimensional space. I will use the manifold learning algorithm <em><a href="https://en.wikipedia.org/wiki/Isomap">ISOMAP</a></em> to calculate a low-dimensional representation that preserves the geodesic distances and then cluster.</p>
<p>For the clustering algorithm, I will use <span class="math inline">\(k\)</span>-means. I will also implement my own version of <span class="math inline">\(k\)</span>-means from scratch in <code>numpy</code> to challenge myself.</p>
<section id="spectral-clustering" class="level4">
<h4 class="anchored" data-anchor-id="spectral-clustering">Spectral Clustering</h4>
<p>For spectral clustering I will first convert the adjacency matrix into a Laplacian matrix, which allows use to have at least one eigenvalue equal to zero.</p>
<p>I tried both the standard Laplacian, <span class="math inline">\(L = D - A\)</span>, and the normalized Laplacian, <span class="math inline">\(L = D^{-\frac{1}{2}} \cdot A \cdot D^{-\frac{1}{2}}\)</span>. Here <span class="math inline">\(D\)</span> is the distance matrix which is a diagonal matrix that sums the weights of the adjacency matrix. I found the standard Laplacian worked better for this application.</p>
<p>For clustering, I used classic <span class="math inline">\(k\)</span>-means. I validated the choice of <span class="math inline">\(k\)</span> with visualizations of the elbow diagram, adjacency matrix, MSA sequences, and eigenvectors.</p>
</section>
<section id="clustering-with-isomap" class="level4">
<h4 class="anchored" data-anchor-id="clustering-with-isomap">Clustering with ISOMAP</h4>
<p>For the ISOMAP, I calculated the geodesic distance matrix <span class="math inline">\(D2\)</span> from the graph by finding the shortest path between each pair of sequences. I next center the matrix using the centering matrix <span class="math inline">\(H\)</span>. We then select the <span class="math inline">\(n\)</span> largest eigenvalues to perform multidimensional reductions (MDS), resulting in a reduced representation matrix <span class="math inline">\(Z\)</span>.</p>
<p>I cluster the matrix <span class="math inline">\(Z\)</span> with <span class="math inline">\(k\)</span>-means and then validate the clusters like with spectral clustering.</p>
</section>
</section>
</section>
<section id="the-results" class="level2">
<h2 class="anchored" data-anchor-id="the-results">The Results</h2>
<p>Both approaches find 5 major strains of SARS-CoV-2 sequences. As we increase the <span class="math inline">\(k\)</span> value, the major strains clusters get subdivided into outlier sequences and sub-strain clusters.</p>
<div id="fig-elbows" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-elbows-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-elbows" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-elb-spec" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-width="params.img_max_width">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-elb-spec-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/elbow_spectral.png" class="lightbox" data-gallery="fig-elbows" title="Figure&nbsp;2&nbsp;(a): Spectral Clustering"><img src="../images/projects/csc/elbow_spectral.png" class="img-fluid figure-img" data-max-width="params.img_max_width" data-ref-parent="fig-elbows"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-elb-spec-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Spectral Clustering
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-elbows" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-elb-iso" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-width="params.img_max_width">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-elb-iso-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/elbow_isomap.png" class="lightbox" data-gallery="fig-elbows" title="Figure&nbsp;2&nbsp;(b): ISOMAP"><img src="../images/projects/csc/elbow_isomap.png" class="img-fluid figure-img" data-max-width="params.img_max_width" data-ref-parent="fig-elbows"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-elb-iso-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) ISOMAP
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-elbows-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Elbow diagrams for spectral clustering and ISOMAP. Both show a break at 5 <span class="math inline">\(k\)</span>=5.
</figcaption>
</figure>
</div>
<section id="spectral-clustering-1" class="level3">
<h3 class="anchored" data-anchor-id="spectral-clustering-1">Spectral Clustering</h3>
<p>If we look at the adjacency matrix for spectral clustering, we can see the effects of the number of clusters. It make five clusters well. Going beyond five clusters, we see the most heterogenous cluster breaking into into sub-clusters.</p>
<div id="fig-adj-mtx" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-adj-mtx-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-adj-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj1" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_1.png" class="lightbox" data-gallery="fig-adj-mtx" title="Figure&nbsp;3&nbsp;(a): k=1"><img src="../images/projects/csc/adj_sorted_1.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-adj-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) <span class="math inline">\(k\)</span>=1
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-adj-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj5" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_5.png" class="lightbox" data-gallery="fig-adj-mtx" title="Figure&nbsp;3&nbsp;(b): k=5"><img src="../images/projects/csc/adj_sorted_5.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-adj-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) <span class="math inline">\(k\)</span>=5
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-adj-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj7" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_7.png" class="lightbox" data-gallery="fig-adj-mtx" title="Figure&nbsp;3&nbsp;(c): k=7"><img src="../images/projects/csc/adj_sorted_7.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-adj-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) <span class="math inline">\(k\)</span>=7
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-adj-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj9" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_9.png" class="lightbox" data-gallery="fig-adj-mtx" title="Figure&nbsp;3&nbsp;(d): k=9"><img src="../images/projects/csc/adj_sorted_9.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-adj-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(d) <span class="math inline">\(k\)</span>=9
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-adj-mtx-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Adjacency matrix after spectral clustering for <span class="math inline">\(k\)</span> clusters.
</figcaption>
</figure>
</div>
<p>Looking at the sequences, we again see that there are 5 major strains, with each strain having different levels of heterogeneity.</p>
<div id="fig-spec-seq" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-spec-seq-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-spec-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj1" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_1.png" class="lightbox" data-gallery="fig-spec-seq" title="Figure&nbsp;4&nbsp;(a): k=1"><img src="../images/projects/csc/sequences_sorted_1.png" data-ref-parent="fig-spec-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) <span class="math inline">\(k\)</span>=1
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-spec-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj5" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_5.png" class="lightbox" data-gallery="fig-spec-seq" title="Figure&nbsp;4&nbsp;(b): k=5"><img src="../images/projects/csc/sequences_sorted_5.png" data-ref-parent="fig-spec-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) <span class="math inline">\(k\)</span>=5
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-spec-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj7" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_7.png" class="lightbox" data-gallery="fig-spec-seq" title="Figure&nbsp;4&nbsp;(c): k=7"><img src="../images/projects/csc/sequences_sorted_7.png" data-ref-parent="fig-spec-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) <span class="math inline">\(k\)</span>=7
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-spec-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj9" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_9.png" class="lightbox" data-gallery="fig-spec-seq" title="Figure&nbsp;4&nbsp;(d): k=9"><img src="../images/projects/csc/sequences_sorted_9.png" data-ref-parent="fig-spec-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(d) <span class="math inline">\(k\)</span>=9
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-spec-seq-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: MSA sequences after sorting for the <span class="math inline">\(k\)</span> spectral clusters.
</figcaption>
</figure>
</div>
<p>We can also look at the clusters in the reduced spectral space. Eigenvectors 1 and 2 show good separation among the 5 clusters. On eigenvector 4, we see separation of one cluster into sub-clusters.</p>
<div id="fig-spec-dims" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-spec-dims-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-spec-dims" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-elb-spec" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-width="params.img_max_width">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-elb-spec-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/spectral_eigs_1_2.png" class="lightbox" data-gallery="fig-spec-dims" title="Figure&nbsp;5&nbsp;(a): Eigenvectors 1 &amp; 2"><img src="../images/projects/csc/spectral_eigs_1_2.png" class="img-fluid figure-img" data-max-width="params.img_max_width" data-ref-parent="fig-spec-dims"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-elb-spec-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Eigenvectors 1 &amp; 2
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-spec-dims" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-elb-iso" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-width="params.img_max_width">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-elb-iso-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/spectral_eigs_4_2.png" class="lightbox" data-gallery="fig-spec-dims" title="Figure&nbsp;5&nbsp;(b): Eigenvectors 4 &amp; 2"><img src="../images/projects/csc/spectral_eigs_4_2.png" class="img-fluid figure-img" data-max-width="params.img_max_width" data-ref-parent="fig-spec-dims"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-elb-iso-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Eigenvectors 4 &amp; 2
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-spec-dims-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Visualization of the eigenvectors. Color coded by cluster where <span class="math inline">\(k=7\)</span>. Note the subdivision of sub-clusters on the left of eigenvector 4.
</figcaption>
</figure>
</div>
</section>
<section id="isomap" class="level3">
<h3 class="anchored" data-anchor-id="isomap">ISOMAP</h3>
<p>The ISOMAP results are similar to the spectral clustering result. We see 5 major strains, with each strain having different levels of heterogeneity. At <span class="math inline">\(k\)</span>=9, the ISOMAP does a better job of separating the sub-strains, as we can see in the geodesic distance matrix below.</p>
<div id="fig-geo-mtx" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-geo-mtx-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-geo-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj1" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_iso_1.png" class="lightbox" data-gallery="fig-geo-mtx" title="Figure&nbsp;6&nbsp;(a): k=1"><img src="../images/projects/csc/adj_sorted_iso_1.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-geo-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) <span class="math inline">\(k\)</span>=1
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-geo-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj5" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_iso_5.png" class="lightbox" data-gallery="fig-geo-mtx" title="Figure&nbsp;6&nbsp;(b): k=5"><img src="../images/projects/csc/adj_sorted_iso_5.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-geo-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) <span class="math inline">\(k\)</span>=5
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-geo-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj7" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_iso_7.png" class="lightbox" data-gallery="fig-geo-mtx" title="Figure&nbsp;6&nbsp;(c): k=7"><img src="../images/projects/csc/adj_sorted_iso_7.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-geo-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) <span class="math inline">\(k\)</span>=7
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-geo-mtx" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj9" class="lightbox quarto-float quarto-figure quarto-figure-center anchored" data-max-with="200px">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/adj_sorted_iso_9.png" class="lightbox" data-gallery="fig-geo-mtx" title="Figure&nbsp;6&nbsp;(d): k=9"><img src="../images/projects/csc/adj_sorted_iso_9.png" class="img-fluid figure-img" data-max-with="200px" data-ref-parent="fig-geo-mtx"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(d) <span class="math inline">\(k\)</span>=9
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-geo-mtx-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Geodesic distance matrix after ISOMAP for <span class="math inline">\(k\)</span> clusters.
</figcaption>
</figure>
</div>
<p>Next we can see the clustering of the sequences.</p>
<div id="fig-iso-seq" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-iso-seq-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-iso-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj1" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_iso_1.png" class="lightbox" data-gallery="fig-iso-seq" title="Figure&nbsp;7&nbsp;(a): k=1"><img src="../images/projects/csc/sequences_sorted_iso_1.png" data-ref-parent="fig-iso-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) <span class="math inline">\(k\)</span>=1
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-iso-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj5" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_iso_5.png" class="lightbox" data-gallery="fig-iso-seq" title="Figure&nbsp;7&nbsp;(b): k=5"><img src="../images/projects/csc/sequences_sorted_iso_5.png" data-ref-parent="fig-iso-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) <span class="math inline">\(k\)</span>=5
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-iso-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj7" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_iso_7.png" class="lightbox" data-gallery="fig-iso-seq" title="Figure&nbsp;7&nbsp;(c): k=7"><img src="../images/projects/csc/sequences_sorted_iso_7.png" data-ref-parent="fig-iso-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) <span class="math inline">\(k\)</span>=7
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-iso-seq" style="flex-basis: 25.0%;justify-content: flex-start;">
<div id="fig-adj9" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="../images/projects/csc/sequences_sorted_iso_9.png" class="lightbox" data-gallery="fig-iso-seq" title="Figure&nbsp;7&nbsp;(d): k=9"><img src="../images/projects/csc/sequences_sorted_iso_9.png" data-ref-parent="fig-iso-seq" height="400" class="figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-adj9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(d) <span class="math inline">\(k\)</span>=9
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-iso-seq-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: MSA sequences after sorting for the <span class="math inline">\(k\)</span> ISOMAP clusters.
</figcaption>
</figure>
</div>
<p>We can visualize the clusters in the reduced ISOMAP space. We see good separation among the 5 clusters, with further separation of some sub-clusters when <span class="math inline">\(k\)</span>=7.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="../images/projects/csc/2D_isomap_7.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26" title="Visualization of the ISOMAP reduced dimensions. Color coded by cluster where k=7."><img src="../images/projects/csc/2D_isomap_7.png" class="img-fluid figure-img" data-max-width="params.img_max_width" alt="Visualization of the ISOMAP reduced dimensions. Color coded by cluster where k=7."></a></p>
<figcaption>Visualization of the ISOMAP reduced dimensions. Color coded by cluster where <span class="math inline">\(k=7\)</span>.</figcaption>
</figure>
</div>
</section>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>Both the spectral clustering and ISOMAP approaches worked well for clustering the SARS-CoV-2 sequences. Both approaches found 5 major strains, with each strain having different levels of heterogeneity. The ISOMAP approach did a better job of separating the sub-strains. Further validation could be performed by using labeled data with strain information.</p>
<p>Further work could be done to identify the sensitive of the results to the data cleaning thresholds. Also, the entropy cutoff could be optimized by fitting a mixture model to the entropy distribution. Finally, the clustering algorithms could be optimized for speed and memory usage.</p>
<p>The clustering of the strains also raises questions of phylogeny. The clustering could be used to build a phylogenetic tree of the strains. This could be done by using the Hamming distance as a measure of genetic distance between the strains. The tree could be built using a variety of phylogenetic tree algorithms, such as UPGMA or neighbor joining. Or we could drop clustering altogether and model the viral evolution as part of a Ornstein-Uhlenbeck process with stochastic differential equations.</p>
<p>I hope this project has given you a good overview of how to cluster cDNA sequences. If you have any questions or suggestions, please feel free to reach out to me.</p>


</section>

</main> <!-- /main -->
<footer style="text-align: center; padding: 10px 0; font-size: 0.8em; color: #666;">
    Â© <span id="current-year"></span> Joshua Meehl. All rights reserved.  Content licensed under
    <a href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank"> CC BY-NC-ND 4.0.
</a></footer><a href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank">
<script>
    const baseYear = 2024;
    const currentYear = new Date().getFullYear();
    const displayYear = currentYear > baseYear ? `${baseYear}-${currentYear}` : `${baseYear}`;
    document.getElementById('current-year').textContent = displayYear;
</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</a></div><a href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank"> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","openEffect":"zoom","selector":".lightbox","loop":false,"descPosition":"bottom"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</a></body></html>